pkg/app/piped/sourceprocesser/sourceprocesser.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2024 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sourceprocesser
+
+import (
+	"fmt"
+)
+
+type SourceTemplateProcessor interface {
+	// BuildTemplateData returns the data that will be used to template target files
+	BuildTemplateData(appDir string) (map[string]string, error)
+	// TemplateSource performs templating prepared data to the source files
+	TemplateSource(appDir string, data map[string](map[string]string)) error
+	// TemplateKey returns the key that will be used to store the data in the template
+	TemplateKey() string
+}
+
+type SourceProcessor interface {
+	Process() error
+}
+
+type processor struct {
+	appDir             string
+	templateProcessors []SourceTemplateProcessor
+}
+
+func NewSourceProcessor(appDir string, templateProcessors ...SourceTemplateProcessor) SourceProcessor {
+	return &processor{
+		appDir:             appDir,
+		templateProcessors: templateProcessors,
+	}
+}
+
+func (p *processor) Process() error {
+	if len(p.templateProcessors) == 0 {
+		return fmt.Errorf("no template processor was specified")
+	}
+
+	// Build the initial data for the template.
+	data := make(map[string](map[string]string))
+	for _, tp := range p.templateProcessors {
+		pdata, err := tp.BuildTemplateData(p.appDir)
+		if err != nil {
+			return err
+		}
+		if len(pdata) == 0 {
+			continue
+		}
+		data[tp.TemplateKey()] = pdata
+	}
+
+	for _, tp := range p.templateProcessors {
+		// Rebuid the data for the template to get up to date
+		// data after previous template processing.
+		pdata, err := tp.BuildTemplateData(p.appDir)
+		if err != nil {
+			return err
+		}
+		if len(pdata) == 0 {
+			continue
+		}
+		data[tp.TemplateKey()] = pdata
+
+		// Perform the templating.
+		if err := tp.TemplateSource(p.appDir, data); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
```

Comment:
Got it! For my understanding, wrote my understanding of the code  (sorry if duplicate former your comment 🙏 )

At first, create values for templating. If the data includes the partial template, treat it as a value to avoid the parse error like "map has no entry for key".

Then iterate in the order below
- embed values to the templates
- then rebuild values to get up-to-date data after previous template processing

Reply:
[nit]
it would be better to execute `tp.TemplateSource` first, then execute `tp.BuildTemplateData` to understand it easily 👀

Reply:
[nit] Refer to the background with comments because this is tricky. 👀 

example↓
- We want to multiple processors step by step, but currently, text/template can't realize it by default (occurs the missing key error).
- So then create values for templating. If the data includes the partial template, treat it as a value.

📝 found the go issue about missing key error
ref: https://github.com/golang/go/issues/60073

Also, later it might be better to explain on docs because the behavior is a little different from the default go template's one 👀 (it's ok on another PR)

Reply:
> [nit]
> 
> it would be better to execute `tp.TemplateSource` first, then execute `tp.BuildTemplateData` to understand it easily 👀 
> 
> 

I don't think so. Build/prepare data then use that data to templating file is more natural to me. wdyt?

Reply:
Also I don't think docs is needed here, this is implementation and tests cover what we mean/expected to be run already. And I can't image how can I write some docs for this too 😅

Reply:
> I don't think so. Build/prepare data then use that data to template file is more natural to me. wdyt?

I see, I got your point and agreed 👍

> Also I don't think docs is needed here, this is implementation and tests cover what we mean/expected to be run already. And I can't image how can I write some docs for this too 😅

I see, also agreed! keep in my mind for test first 👍